Merge run provider tests and use built provider (#8229)
GoogleCloudPlatform/magic-modules: docs/content/develop/run-tests.md

    ```bash
    make testacc TEST=./google-beta TESTARGS='-run=TestAccContainerNodePool'
    ```

1. Optional: Save verbose test output to a file for analysis.

    ```bash
    TF_LOG=TRACE make testacc TEST=./google-beta TESTARGS='-run=TestAccContainerNodePool_basic' > output.log
    ```

1. Optional: Debug tests with [Delve](https://github.com/go-delve/delve). See [`dlv test` documentation](https://github.com/go-delve/delve/blob/master/Documentation/usage/dlv_test.md) for information about available flags.

    ```bash
    cd google-beta
    TF_ACC=1 dlv test -- --test.v --test.run TestAccComputeRegionBackendService_withCdnPolicy
    ```

{{< /tab >}}

{{< /tabs >}}

## Test with different `terraform` versions

Tests will use whatever version of the `terraform` binary is found on your `PATH`. If you are testing a change that you know only impacts certain `terraform` versions, follow these steps:

1. Install [`tfenv`](https://github.com/tfutils/tfenv).

1. Install the version of `terraform` you want to test.

    ```bash
    tfenv install VERSION
    ```

    Replace `VERSION` with the version you want to test.

> 1. Run automated tests following the earlier section.

Reviewer: please add a link to the earlier section here & in the next item

Author: I've removed the next item because all the tests are now under the automated tests section.

make build
```

{{< tabs "built-provider" >}}

{{< tab "Developer overrides" >}}

In the sections below we describe how to create a Terraform CLI configuration file, and how to make the CLI use the file via an environment variable.

### Create developer overrides file

Choose your architecture below.

{{< tabs "setup" >}}

{{< tab "Mac (ARM64 and AMD64), Linux AMD64" >}}

1. Find the location where built provider binaries are created. To do this, run this command and make a note of the path value:

    ```bash
    go env GOBIN

    ## If the above returns nothing, then run the command below and add "/bin" to the end of the output path.
    go env GOPATH
    ```

1. Create an empty configuration file.

    ```bash
    ## create an empty file
    touch ~/tf-dev-override.tfrc

    ## open the file with a text editor of your choice, e.g:
    vi ~/tf-dev-override.tfrc
    ```

    Open the empty file with a text editor and paste in these contents:

    ```hcl
    provider_installation {

      # Developer overrides will stop Terraform from downloading the listed
      # providers their origin provider registries.
      dev_overrides {

>           "hashicorp/google" = "<REPLACE-ME>/bin"

Reviewer: see https://developers.google.com/style/placeholders for style guidance - should be a meaningful value, all caps, with underscores.

---
title: "Run tests"
weight: 26
aliases:
  - /docs/getting-started/run-provider-tests
  - /docs/getting-started/use-built-provider
  - /docs/how-to/run-tests
  - /get-started/run-provider-tests
  - /get-started/use-built-provider
  - /getting-started/run-provider-tests
  - /getting-started/use-built-provider
  - /how-to/run-tests
---

> # Run tests

Author: Done.

## Run provider tests locally

{{< hint info >}}

> **Note:** If you want to test changes you've made in Magic Modules, you need to first [generate](/magic-modules/docs/getting-started/generate-providers/) the provider you want to test.

Author: Done.

{{< /hint >}}

### Setup

> Authentication is described in more detail [here](https://github.com/hashicorp/terraform-provider-google/wiki/Developer-Best-Practices#authentication).

Author: Done.

In order to run tests, set the following environment variables:

```

> GOOGLE_PROJECT

Author: Done.

```bash
## for ga provider
cd $GOPATH/src/github.com/hashicorp/terraform-provider-google
make testacc TEST=./google TESTARGS='-run=TestAccContainerNodePool_basic'

## for beta provider
cd $GOPATH/src/github.com/hashicorp/terraform-provider-google-beta
make testacc TEST=./google-beta TESTARGS='-run=TestAccContainerNodePool_basic'
```

TESTARGS allows you to pass [testing flags](https://pkg.go.dev/cmd/go#hdr-Testing_flags) to `go test`. The most important is `-run`, which allows you to limit the tests that get run. There are 2000+ tests, and running all of them takes over 9 hours and requires a lot of GCP quota.

`-run` is regexp-like, so multiple tests can be run in parallel by specifying a common substring of those tests (for example, `TestAccContainerNodePool` to run all node pool tests).

### Debugging tests

You can [increase your test verbosity](https://www.terraform.io/docs/internals/debugging.html)  and redirect the output to a log file for analysis. This is often helpful in debugging issues.

```bash
## for ga provider
cd $GOPATH/src/github.com/hashicorp/terraform-provider-google

> TF_LOG=TRACE make testacc TEST=./google TESTARGS='-run=TestAccContainerNodePool_basic' > output.log

Author: This is now a sub-heading of "Run acceptance tests".

## for beta provider
cd $GOPATH/src/github.com/hashicorp/terraform-provider-google-beta
TF_LOG=TRACE make testacc TEST=./google-beta TESTARGS='-run=TestAccContainerNodePool_basic' > output.log
```

You can also debug tests with [Delve](https://github.com/go-delve/delve):

```bash
## Navigate to the google package within your local GCP Terraform provider Git clone.
cd $GOPATH/src/github.com/terraform-providers/terraform-provider-google/google

## Execute the dlv command to launch the test.
## Note that the --test.run flag uses the same regexp matching as go test --run.

> TF_ACC=1 dlv test -- --test.v --test.run TestAccComputeRegionBackendService_withCdnPolicy

Author: Done.

   535:                         },
   536:                 },
   537:         })
   538: }
   539:
=> 540: func TestAccComputeRegionBackendService_withCdnPolicy(t *testing.T) {
   541:         t.Parallel()
   542:
   543:         var svc compute.BackendService
   544:         resource.Test(t, resource.TestCase{
   545:                 PreCheck:     func() { acctest.AccTestPreCheck(t) },
(dlv)
```

### Testing with different `terraform` versions

Tests will use whatever version of the `terraform` binary is found on your path. To test with multiple versions of `terraform` core, you must run the tests multiple times with different versions. You can use [`tfenv`](https://github.com/tfutils/tfenv) to manage your system `terraform` versions.

## Use built provider locally

Sometimes, for example for manual testing, you may want to build the provider from source and use it with `terraform`.

### Developer Overrides

{{< hint info >}}
**Note:** Developer overrides are only available in Terraform v0.14 and later.
{{< /hint >}}

In the sections below we describe how to create a Terraform CLI configuration file, and how to make the CLI use the file via an environment variable.

### Setup

> Choose your architecture below.

Author: Done.

---
title: "Run tests"
weight: 70
aliases:
  - /docs/getting-started/run-provider-tests
  - /docs/getting-started/use-built-provider
  - /get-started/run-provider-tests
  - /get-started/use-built-provider
  - /getting-started/run-provider-tests
  - /getting-started/use-built-provider
---

# Run tests

## Before you begin

[Generate](/magic-modules/docs/getting-started/generate-providers/) the provider you want to test.

We reccomend using using "application default credentials" and running tests as your GCP user account.

1. Log in with gcloud:

    ```bash
    gcloud auth application-default login
    ```

1. Set the following environment variables:

    ```bash
    export GOOGLE_USE_DEFAULT_CREDENTIALS=TRUE
    export GOOGLE_PROJECT=PROJECT_ID
    export GOOGLE_REGION=us-central1
    export GOOGLE_ZONE=us-central1-a
    ```

> Additional variables may be required for other tests, and should get flagged when running them by Go skipping the test and flagging in the output it was skipped, with a skip message explaining why. The most typical extra values required are those required for project creation:

Author: Done.

```
GOOGLE_ORG
GOOGLE_BILLING_ACCOUNT
```

## Run unit tests

Unit tests (that is, tests that do not interact with the GCP API) are very fast and you can generally run them all if you have changed any of them:

```bash

> ## for ga provider

Author: Done.

cd $GOPATH/src/github.com/hashicorp/terraform-provider-google
make test
make lint

## for beta provider
cd $GOPATH/src/github.com/hashicorp/terraform-provider-google-beta
make test
make lint
```

## Run acceptance tests

You can run tests against the provider you generated in the `OUTPUT_PATH` location. When running tests, specify which to run using `TESTARGS`, such as:

```bash
## for ga provider

> cd $GOPATH/src/github.com/hashicorp/terraform-provider-google

Author: Done.

make testacc TEST=./google-beta TESTARGS='-run=TestAccContainerNodePool_basic'
```

TESTARGS allows you to pass [testing flags](https://pkg.go.dev/cmd/go#hdr-Testing_flags) to `go test`. The most important is `-run`, which allows you to limit the tests that get run. There are 2000+ tests, and running all of them takes over 9 hours and requires a lot of GCP quota.

`-run` is regexp-like, so multiple tests can be run in parallel by specifying a common substring of those tests (for example, `TestAccContainerNodePool` to run all node pool tests).

### Debugging tests

You can [increase your test verbosity](https://www.terraform.io/docs/internals/debugging.html) and redirect the output to a log file for analysis. This is often helpful in debugging issues.

```bash
## for ga provider
cd $GOPATH/src/github.com/hashicorp/terraform-provider-google
TF_LOG=TRACE make testacc TEST=./google TESTARGS='-run=TestAccContainerNodePool_basic' > output.log

## for beta provider
cd $GOPATH/src/github.com/hashicorp/terraform-provider-google-beta
TF_LOG=TRACE make testacc TEST=./google-beta TESTARGS='-run=TestAccContainerNodePool_basic' > output.log
```

> You can also debug tests with [Delve](https://github.com/go-delve/delve):

Author: Done.

```bash
## Navigate to the google package within your local GCP Terraform provider Git clone.
cd $GOPATH/src/github.com/terraform-providers/terraform-provider-google/google

## Execute the dlv command to launch the test.
## Note that the --test.run flag uses the same regexp matching as go test --run.
TF_ACC=1 dlv test -- --test.v --test.run TestAccComputeRegionBackendService_withCdnPolicy
```

### Testing with different `terraform` versions

> Tests will use whatever version of the `terraform` binary is found on your path. To test with multiple versions of `terraform` core, you must run the tests multiple times with different versions. You can use [`tfenv`](https://github.com/tfutils/tfenv) to manage your system `terraform` versions.

Reviewer: Possibly something like:

```suggestion
Tests will use whatever version of the `terraform` binary is found on your `PATH`. If you are testing a change that you know only impacts certain `terraform` versions, follow these steps:

1. Install [`tfenv`](https://github.com/tfutils/tfenv).
2. Install the version of `terraform` you want to test.
   \`\`\`bash
   tfenv install VERSION
   \`\`\`
   Replace `VERSION` with the version you want to test.
3. Run automated tests following the earlier section.
4. Run manual tests following the earlier section.
```

(Can't nest code blocks apparently :joy:)

## Use built provider locally

Sometimes, for example for manual testing, you may want to build the provider from source and use it with `terraform`.

### Build provider

```bash
## ga provider
cd $GOPATH/src/github.com/hashicorp/terraform-provider-google
make build

## beta provider
cd $GOPATH/src/github.com/hashicorp/terraform-provider-google-beta
make build
```

{{< tabs "built-provider" >}}

{{< tab "Developer overrides" >}}

{{< hint info >}}

> **Note:** Developer overrides are only available in Terraform v0.14 and later.

Author: Done.

---
title: "Run tests"
weight: 70
aliases:
  - /docs/getting-started/run-provider-tests
  - /docs/getting-started/use-built-provider
  - /get-started/run-provider-tests
  - /get-started/use-built-provider
  - /getting-started/run-provider-tests
  - /getting-started/use-built-provider
---

# Run tests

## Before you begin

1. [Generate the modified provider(s)](/magic-modules/docs/getting-started/generate-providers/)

1. Set up application default credentials for Terraform

    ```bash
    gcloud auth application-default login
    export GOOGLE_USE_DEFAULT_CREDENTIALS=TRUE
    ```

1. Set the following environment variables:

    ```bash
    export GOOGLE_USE_DEFAULT_CREDENTIALS=TRUE
    export GOOGLE_PROJECT=PROJECT_ID
    export GOOGLE_REGION=us-central1
    export GOOGLE_ZONE=us-central1-a
    ```

1. Optional: Some tests may require additional variables to be set, such as:

    ```
    GOOGLE_ORG
    GOOGLE_BILLING_ACCOUNT
    ```

{{< tabs "version" >}}

{{< tab "GA Provider" >}}

## Run automated tests

Run unit tests and linters

```bash
cd $GOPATH/src/github.com/hashicorp/terraform-provider-google
make test
make lint
```

> ## Run acceptance tests

Reviewer: ```suggestion
```

> Run acceptance tests

Reviewer: What would you think about linking to Go's documentation about `-run` instead of having the following two paragraphs?

```suggestion
2. Run acceptance tests for only modified resources. (Full test runs can take over 9 hours.) See Go's documentation for more information about `-run`.
```

```bash

> make testacc TEST=./google TESTARGS='-run=TestAccContainerNodePool_basic'

Reviewer: ```suggestion
make testacc TEST=./google TESTARGS='-run=TestAccContainerNodePool'
```

```bash
TF_LOG=TRACE make testacc TEST=./google TESTARGS='-run=TestAccContainerNodePool_basic' > output.log
```

{{< /tab >}}

{{< tab "Beta Provider" >}}

## Run automated tests

Run unit tests and linters

```bash
cd $GOPATH/src/github.com/hashicorp/terraform-provider-google-beta
make test
make lint
```

## Run acceptance tests

Run acceptance tests

```bash
make testacc TEST=./google-beta TESTARGS='-run=TestAccContainerNodePool_basic'
```

TESTARGS allows you to pass [testing flags](https://pkg.go.dev/cmd/go#hdr-Testing_flags) to `go test`. The most important is `-run`, which allows you to limit the tests that get run. There are 2000+ tests, and running all of them takes over 9 hours and requires a lot of GCP quota.

`-run` is regexp-like, so multiple tests can be run in parallel by specifying a common substring of those tests (for example, `TestAccContainerNodePool` to run all node pool tests).

### Debugging tests

Optional: Save verbose test output to a file for analysis.

```bash
TF_LOG=TRACE make testacc TEST=./google-beta TESTARGS='-run=TestAccContainerNodePool_basic' > output.log
```

{{< /tab >}}

{{< /tabs >}}

> You can also debug tests with [Delve](https://github.com/go-delve/delve):

Reviewer: Could this be phrased as an optional step in the tabs above?

Author: Done.

  - /docs/getting-started/use-built-provider
  - /get-started/run-provider-tests
  - /get-started/use-built-provider
  - /getting-started/run-provider-tests
  - /getting-started/use-built-provider
---

# Run tests

## Before you begin

1. [Generate the modified provider(s)](/magic-modules/docs/getting-started/generate-providers/)

1. Set up application default credentials for Terraform

    ```bash
    gcloud auth application-default login
    export GOOGLE_USE_DEFAULT_CREDENTIALS=TRUE
    ```

1. Set the following environment variables:

    ```bash
    export GOOGLE_USE_DEFAULT_CREDENTIALS=TRUE
    export GOOGLE_PROJECT=PROJECT_ID
    export GOOGLE_REGION=us-central1
    export GOOGLE_ZONE=us-central1-a
    ```
    Replace `PROJECT_ID` with the ID of the Google Cloud project you are using for testing.

1. Optional: Some tests may require additional variables to be set, such as:

    ```
    GOOGLE_ORG
    GOOGLE_BILLING_ACCOUNT
    ```

## Run automated tests

{{< tabs "version" >}}

{{< tab "GA Provider" >}}

1. Run unit tests and linters

    ```bash
    cd $GOPATH/src/github.com/hashicorp/terraform-provider-google
    make test
    make lint
    ```

> 1. Run acceptance tests for only modified resources. (Full test runs can take over 9 hours.) See Go's documentation for more information about `-run`.

Reviewer: ```suggestion
1. Run acceptance tests for only modified resources. (Full test runs can take over 9 hours.) See [Go's documentation](https://pkg.go.dev/cmd/go#hdr-Testing_flags) for more information about `-run` and other flags.
```

    ```bash
    make testacc TEST=./google TESTARGS='-run=TestAccContainerNodePool'
    ```

>     TESTARGS allows you to pass [testing flags](https://pkg.go.dev/cmd/go#hdr-Testing_flags) to `go test`. The most important is `-run`, which allows you to limit the tests that get run. There are 2000+ tests, and running all of them takes over 9 hours and requires a lot of GCP quota.

Reviewer: ```suggestion
```

>     `-run` is regexp-like, so multiple tests can be run in parallel by specifying a common substring of those tests (for example, `TestAccContainerNodePool` to run all node pool tests).

Reviewer: ```suggestion
```

1. Optional: Save verbose test output to a file for analysis.

    ```bash
    TF_LOG=TRACE make testacc TEST=./google TESTARGS='-run=TestAccContainerNodePool_basic' > output.log
    ```

> 1. Optional: debug tests with [Delve](https://github.com/go-delve/delve):

Reviewer: ```suggestion
1. Optional: Debug tests with [Delve](https://github.com/go-delve/delve). See [`dlv test` documentation](https://github.com/go-delve/delve/blob/master/Documentation/usage/dlv_test.md) for information about available flags.
```

    ```bash
    ## Navigate to the google package within your local GCP Terraform provider Git clone.

>     cd $GOPATH/src/github.com/terraform-providers/terraform-provider-google/google

Reviewer: This line shouldn't be necessary since they're already in the directory in this part of the guide.

```suggestion
```

    ## Execute the dlv command to launch the test.

>     ## Note that the --test.run flag uses the same regexp matching as go test --run.

Reviewer: These comments can probably be removed

    TF_ACC=1 dlv test -- --test.v --test.run TestAccComputeRegionBackendService_withCdnPolicy
    ```

{{< /tab >}}

> {{< tab "Beta Provider" >}}

Reviewer: Please update this section to match the GA Provider section

Reviewer: this is done

    `-run` is regexp-like, so multiple tests can be run in parallel by specifying a common substring of those tests (for example, `TestAccContainerNodePool` to run all node pool tests).

1. Optional: Save verbose test output to a file for analysis.

    ```bash
    TF_LOG=TRACE make testacc TEST=./google-beta TESTARGS='-run=TestAccContainerNodePool_basic' > output.log
    ```

1. Optional: debug tests with [Delve](https://github.com/go-delve/delve):

    ```bash
    ## Navigate to the google package within your local GCP Terraform provider Git clone.
    cd $GOPATH/src/github.com/terraform-providers/terraform-provider-google-beta/google-beta

    ## Execute the dlv command to launch the test.
    ## Note that the --test.run flag uses the same regexp matching as go test --run.
    TF_ACC=1 dlv test -- --test.v --test.run TestAccComputeRegionBackendService_withCdnPolicy
    ```

{{< /tab >}}

{{< /tabs >}}

> ### Testing with different `terraform` versions

Reviewer: Looking at the current state of the doc, I think this could probably be moved to the end and made a sibling of the other sections.

```suggestion
## Test with different `terraform` versions
```

Tests will use whatever version of the `terraform` binary is found on your path. To test with multiple versions of `terraform` core, you must run the tests multiple times with different versions. You can use [`tfenv`](https://github.com/tfutils/tfenv) to manage your system `terraform` versions.

> ## Test manually

Reviewer: Please rewrite this section to be more procedural and provide less context. Possibly something like the following subsections:

- Set up test environment
   - Could include setting up developer overrides or filesystem mirrors
- Run tests
  - Could include building the provider, creating a test directory and test configurations, and running `terraform apply` with debug output
- Cleanup
   - Could include any teardown for developer overrides or filesystem mirrors.

For filesystem mirrors - like we talked about, it's okay to leave that content less procedural for now since it's not the default path. However, developer overrides should be procedural, with less context overall - in particular, we should only recommend one path for using them, either global configuration or an environment variable. (I'd have a slight preference for an environment variable that they pass directly in copy-pasteable commands but am open to other options as long as there is a single recommended path.)

Reviewer: ```suggestion
## Optional: Test manually
```

The developer overrides section is looking better, but it would be great to format this in sections like:

- Setup
- Run manual tests
  - This would include building the provider
- Cleanup

The main changes for this are:

1. Move the provider build step into the "Using Terraform CLI developer overrides" section
2. Update section headers & move them out of the tabs (so they show up in the sidebar nav)
   - This will mean setting up tabs in each section & adding content for the "Run manual tests" section for filesystem mirrors.

Sometimes, for example for manual testing, you may want to build the provider from source and use it with `terraform`.

### Build provider

```bash
## ga provider
cd $GOPATH/src/github.com/hashicorp/terraform-provider-google
make build

## beta provider
cd $GOPATH/src/github.com/hashicorp/terraform-provider-google-beta
make build
```

{{< tabs "built-provider" >}}

> {{< tab "Developer overrides" >}}

Reviewer: ```suggestion
{{< tab "Developer overrides (Terraform 0.14 and later)" >}}
```

This would also allow deleting the note below.

Reviewer: per discussion, leaving this off as it's implied.

{{< hint info >}}
**Note:** Developer overrides are only available in Terraform v0.14 and later.
{{< /hint >}}

In the sections below we describe how to create a Terraform CLI configuration file, and how to make the CLI use the file via an environment variable.

> ### Setup

Reviewer: ```suggestion
### Create developer overrides file
```

This CLI configuration file you created in the steps above will allow Terraform to automatically use the binaries generated by the `make build` commands in the `terraform-provider-google` and `terraform-provider-google-beta` repositories instead of downloading the latest versions. All other providers will continue to be downloaded from the public Registry as normal.

> ### Build provider

Author: I think this would be more confusing since this step is only required for manual testing.

**Note:** Developer overrides work without you needing to alter your Terraform configuration in any way.
{{< /hint >}}

> ### Download production providers

Author: Done.

{{< /hint >}}

### Download production providers

To stop using developer overrides, unset the `TF_CLI_CONFIG_FILE` environment variable or stop setting it in the commands you are executing.

This will then let Terraform resume normal behaviour of pulling published provider versions from the public Registry. Any version constraints in your Terraform configuration will come back into effect. Also, you may need to run `terraform init` to download the required version of the provider into your project directory if you haven't already.

### Alternative: using a global CLI configuration file

If you do not want to use the `TF_CLI_CONFIG_FILE` environment variable, as described above, you can instead create a global version of the CLI configuration file. This configuration will be used automatically by Terraform. To do this, follow [the guidance in the official documentation](https://developer.hashicorp.com/terraform/cli/config/config-file#locations).

In this scenario you will need to remember to edit this file to swap between using developer overrides and using the production provider versions.

> ### Possible problems

Author: Done.

{{< /hint >}}

### Download production providers

To stop using developer overrides, unset the `TF_CLI_CONFIG_FILE` environment variable or stop setting it in the commands you are executing.

This will then let Terraform resume normal behaviour of pulling published provider versions from the public Registry. Any version constraints in your Terraform configuration will come back into effect. Also, you may need to run `terraform init` to download the required version of the provider into your project directory if you haven't already.

### Alternative: using a global CLI configuration file

If you do not want to use the `TF_CLI_CONFIG_FILE` environment variable, as described above, you can instead create a global version of the CLI configuration file. This configuration will be used automatically by Terraform. To do this, follow [the guidance in the official documentation](https://developer.hashicorp.com/terraform/cli/config/config-file#locations).

In this scenario you will need to remember to edit this file to swap between using developer overrides and using the production provider versions.

{{< /tab >}}

> {{< tab "Filesystem mirrors" >}}

Reviewer: ```suggestion
{{< tab "Filesystem mirrors (Terraform 0.13 and earlier)" >}}
```
